src/declaration.c

```diff
@@ -2063,36 +2063,42 @@ Expression *VarDeclaration::callScopeDtor(Scope *sc)
     Type *tv = type->baseElemOf();
     if (tv->ty == Tstruct)
     {
-        TypeStruct *ts = (TypeStruct *)tv;
-        StructDeclaration *sd = ts->sym;
-        if (sd->dtor)
+        StructDeclaration *sd = ((TypeStruct *)tv)->sym;
+        if (!sd->dtor || !type->size())
+            return NULL;
+
+        if (type->toBasetype()->ty == Tstruct)
         {
-            if (type->toBasetype()->ty == Tsarray)
-            {
-                // Typeinfo.destroy(cast(void*)&v);
-                Expression *ea = new SymOffExp(loc, this, 0, 0);
-                ea = new CastExp(loc, ea, Type::tvoid->pointerTo());
-                Expressions *args = new Expressions();
-                args->push(ea);
+            // v.__xdtor()
+            e = new VarExp(loc, this);
 
-                Expression *et = new TypeidExp(loc, type);
-                et = new DotIdExp(loc, et, Id::destroy);
+            /* This is a hack so we can call destructors on const/immutable objects.
+             * Need to add things like "const ~this()" and "immutable ~this()" to
+             * fix properly.
+             */
+            e->type = e->type->mutableOf();
```

Comment:
Because `VarExp` is one of the leaf AST node. Its `semantic` will return itself. So the type setting is low risk.